geopandas/tools/overlay.py

```diff
@@ -75,23 +74,44 @@ def _overlay_difference(df1, df2):
     """
     Overlay Difference operation used in overlay function
     """
-    # Spatial Index to create intersections
-    spatial_index = df2.sindex
-    bbox = df1.geometry.apply(lambda x: x.bounds)
-    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
-    # Create differences
-    new_g = []
-    for geom, neighbours in zip(df1.geometry, sidx):
-        new = reduce(
-            lambda x, y: x.difference(y), [geom] + list(df2.geometry.iloc[neighbours])
+    # spatial index query to find intersections
+    idx1, idx2 = df2.sindex.query_bulk(df1.geometry, predicate="intersects", sort=True)
+    idx1_unique, idx1_unique_idxs = np.unique(idx1, return_index=True)
+    df2_neighbors = (
+        df2.geometry.values.data[idxs] for idxs in np.split(idx2, idx1_unique_idxs[1:])
+    )
+    # copy the input dataframe
+    dfdiff = df1.copy()
+    if compat.USE_PYGEOS:
```

Comment:
Can we leave this vectorisation for a follow-up PR and keep this one focused on `query_bulk` only? 

Just a side note - we have to keep in mind that everything based on pygeos will need to be changed for shapely 2.0 at one moment, so I'd rather avoid situation when we have pygeos-based code scattered around the repo instead of having it at few places (`_vectorized.py`, `sindex.py`) only.

Reply:
Makes sense. How do you feel about keeping the optimization I made for the non pygeos logic (which I think will also help pygeos)?

Reply:
The same. For the review purposes and potential debugging, it is better to make small targeted PRs. It is much easier to review more small changes than one large refactoring. 

You can move both to the same PR focusing on vectorisation of overlay. Once cleaned, this PR can be quickly merged as was `clip`, so it will not block further changes.

Reply:
Okay just pushed a minimal diff version. I do think it is important to at least implement some optimization to this method, it will make a big difference in performance.